Prevent GDScript language server from listening to external hosts by default

Houkime opened the pull request:
* Add "bind_ip" property to WebSocketServer class which defaults to "*" (for compatibility with current users)
    * _WebSocketServer is the class that GDScript language server relies on_
* Add an editor option to set the bind_ip for the GDScript language server
* Set GDscript language server to listen only to localhost by default
_______________________

### Problem

Currently GDScript language server is enabled at startup and exposes an open port (6008 by default) to the net accessible by any host unless filtered.

I feel uncomfortable about it considering that 

* I don't even use the language server function.
* It cannot be turned off and/or bound to localhost from editor settings.

### Language server at startup by default?

The most straightforward solution would be to have the Language Server simply turned off at startup by default.
However at preliminary discussion @Faless at IRC stated that C# users are unlikely to appreciate if it will be turned off by default.
So a solution then to have it on by default, but listen only to localhost, which most likely covers 99% of intended usage anyway.

### Why not just make an off switch, so you, Houkime, turn this off, and others have it on by default?

I consider adding an off switch in a separate PR since it is an independent change.  
This PR however considers the security of those Godot users who are not likely to look at network settings or scan open ports (which is probably a majority) so it is a bit more important to make early.

### Why a property and not an argument for the listen() function?

It was recommended by @Faless for compatibility reasons and because bind_ip is unlikely to change often.

However, in places specific to GDscript Language server (where interfaces are not accessible from GDScript) it is an argument in the line with the underlying TCP_server listen() function. This can be changed however if considered clumsy (though i somehow think that making it a property of a gdscript ptotocol for example is even more odd).

### Why touching WSLServer?

WebSocketServer class is more of a virtual thing, which exact implementation depends on whether we're in browser or running natively (in browser it is a dummy). 
It doesn't implement a listen() function.

ClassDB::instance() when asked for WebSocketServer returns either WSL (native) server or EMWS server.
See https://github.com/godotengine/godot/blob/master/modules/websocket/websocket_server.cpp

akien-mga commented:
Could you squash commits into one?

Houkime commented:
> Could you squash commits into one?

Most probably, though never done this before.
I did check contribution guidelines and pr_worfkflow in documentation beforehead so you might consider adding sth about squashing there if it is sth that is desired often.

akien-mga commented:
See https://docs.godotengine.org/en/3.1/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

Houkime commented:
> See https://docs.godotengine.org/en/3.1/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

Oh, seems like missed it somehow.
Sorry about that and thanks.

Calinou commented:
Is there any use for listening to non-localhost addresses? If not, we should default to listening only on `localhost`. What do other editors do?

Houkime commented:
> Is there any use for listening to non-localhost addresses in the LSP? If not, we should default to listening only on `localhost`.

Theoretically you can have sth like a corporate workspace with a single language server being shared by everyone or even a worldwide-visible language server. 
According to wiki, there was a collaboration effort between Redhat, Microsoft and Eclipsein this direction https://en.wikipedia.org/wiki/Language_Server_Protocol#cite_note-13.
However, I don't have any real examples on that.

I have no real objections to removing the editor option and just hardwiring localhost setting into a language server (better) or a protocol (not really sure). 
Editor option is there primarily because it is easy and doesn't take the risk of ruining someone's existing workflow. I might be overcautious here.

> What do other LSP services do in comparison to ours?

I will check the external host option availability in other LSP servers once I get sleep and unless someone checks it out before me.

Calinou commented:
> I have no real objections to removing the editor option and just hardwiring "localhost" or "127.0.0.1" into a language server (better) or a protocol (not really sure).

Indeed, it seems like this is the way to go for now.

Houkime commented:
> What do other LSP services do in comparison to ours?

Ok, ran a little bit of a research on lsp servers.
In a file is a list of github repos from microsoft language server implemetations list.

[rawlinks.txt](https://github.com/godotengine/godot/files/3994975/rawlinks.txt)

Actual server repos are around 87.
Went through them one by one and checked what kind of (not mutually exclusive) communication they use and what options they have.

Mostly they have stdio communication supported - VScode or whatever else runs the server as a subprocess and communticates via streams.
Then there are HTTP options, with or without changeable port or host.
Mostly HTTP-capable servers can change a port number, but very rarely - host.

In fact, from around 87 I found only the following implementations that are able to change host to listen to:

* https://github.com/camel-tooling/camel-language-server
* https://github.com/eclipse/eclipse.jdt.ls
* https://github.com/felixfbecker/php-language-server
* https://github.com/lingua-pupuli/puppet-editor-services.git

Faless commented on modules/websocket/websocket_server.h:
```diff
@@ -41,6 +41,8 @@ class WebSocketServer : public WebSocketMultiplayerPeer {
 	GDCLASS(WebSocketServer, WebSocketMultiplayerPeer);
 	GDCICLASS(WebSocketServer);
 
+	String bind_ip = "*";
```
Please use `IP_Address` type instead of string.
Also initialize it's value in the constructor, not in the header (for consistency with the rest of the module until we officially move to newer version of C++).

Houkime replied:
Unless i miss sth important IP_Adress class can not be used from GDScript and thus i refrained from using it in order to make host settings changeable from script.
(WebSocketServer itself is usable from GDScript)
Please tell me if i was mistaken in my analysis of the situation.

Houkime replied:
should be fixed now

Faless commented on modules/websocket/websocket_server.h:
```diff
@@ -67,6 +69,9 @@ class WebSocketServer : public WebSocketMultiplayerPeer {
 	void _on_disconnect(int32_t p_peer_id, bool p_was_clean);
 	void _on_close_request(int32_t p_peer_id, int p_code, String p_reason);
 
+	String get_bind_ip() const;
```
Use `IP_Address` instead of `String`.

Houkime replied:
should be fixed now

Faless commented on modules/websocket/websocket_server.h:
```diff
@@ -67,6 +69,9 @@ class WebSocketServer : public WebSocketMultiplayerPeer {
 	void _on_close_request(int32_t p_peer_id, int p_code, String p_reason);
 
+	String get_bind_ip() const;
+	void set_bind_ip(const String &p_bind_ip);
```
Use `IP_Address` instead of `String`.

Houkime replied:
should be fixed now

Faless commented on modules/websocket/wsl_server.cpp:
```diff
@@ -165,7 +165,7 @@ Error WSLServer::listen(int p_port, const Vector<String> p_protocols, bool gd_mp
 		pw[i] = p_protocols[i].strip_edges();
 	}
-	_server->listen(p_port);
+	_server->listen(p_port, IP_Address(bind_ip));
```
`IP_Address(bind_ip)` will be redundant when `bind_ip` is changed to `IP_Address` type

Houkime replied:
should be fixed now

Faless commented on modules/websocket/websocket_server.cpp:
```diff
@@ -67,6 +71,15 @@ void WebSocketServer::_bind_methods() {
+}
+
+void WebSocketServer::set_bind_ip(const String &p_bind_ip) {
+	ERR_FAIL_COND(is_listening());
```
Should also check the validity of the IP address:
`ERR_FAIL_COND(!p_bind_address.is_valid() && !p_bind_address.is_wildcard())`

Houkime replied:
Why disallowing setting a wildcard?

Houkime replied:
added

Faless reviewed (changes requested):
Looking good overall, see my comments

Houkime commented:
@Faless i used the String because IP_Adress class is not exposed to GDScript, while bind_ip probably needs to be accessible from GDScript as the rest of functionality of WebSocketServer.

Faless commented:
> In other words, if i make it an IP_Adress, the user won't be able to change it via script.
> There seems to be no way for a user to even make an IP_Adress in GDscript.

`IP_Address` is automatically casted to a String when passed to GDScript (and vice versa).
Check out `NetworkedMultiplayerENet` for reference. (or many other methods that uses IP_Address, like `PacketPeerUDP::get_packet_address`)

https://github.com/godotengine/godot/blob/a7f49ac9a107820a62677ee3fb49d38982a25165/modules/enet/networked_multiplayer_enet.cpp#L918

 https://github.com/godotengine/godot/blob/a7f49ac9a107820a62677ee3fb49d38982a25165/modules/enet/networked_multiplayer_enet.cpp#L858

Commit:
Prevent GDScript language server from listening to external hosts by default

* Add bind_ip property to WebSocketServer defaulting to "*" (listen to everyone)
* Set default for GDscript Language Server to listen only to localhost

Fixes potential security issue with GDScript language server being exposed to the
broad net by default.
Since it is the server which primary usage is to provide utility to the local
editor there is no need to expose it.

Houkime commented:
@Calinou: removed an editor setting and made GDScriptLanguageServer hardwired to localhost.
@Faless: addressed your requested changes.

Houkime commented:
While GDScript language server gets restricted to localhost and works as intended, due to https://github.com/godotengine/godot/issues/34751 i can't really test out a usage of host restrictions in a more general WebSocketServer case.

Houkime commented:
Ok, with #34751 as it turns out it was my mistake in the end.

Now i was able to test restriction in the general WebSocketServer usage and it seems to be working.

A sample script was like this:

```
extends Node2D

var server

# Called when the node enters the scene tree for the first time.
func _ready():
	server = WebSocketServer.new()
	print("server's bind ip is ", server.bind_ip)
	server.bind_ip = "127.0.0.1"
	print("and now server's bind ip is ", server.bind_ip)
	server.listen(9807)
	print (server.is_listening())

```

(and then check with `lsof -i` that `localhost:9807` indeed appeared)

Faless reviewed (approved):
Looking good to me. Great job :+1: , and sorry about the late reply. Could you rebase to resolve the small merge conflict that arose?

Houkime commented:
> Looking good to me. Great job  , and sorry about the late reply. Could you rebase to resolve the small merge conflict that arose?

rebased

akien-mga commented:
Thanks!
